cpp/src/arrow/compute/kernels/scalar_string.cc

```diff
@@ -48,6 +48,16 @@ struct AsciiUpper {
   }
 };
 
+struct AsciiLower {
+  template <typename... Ignored>
+  static std::string Call(KernelContext*, const util::string_view& val) {
+    std::string result = val.to_string();
```

Comment:
This is completely inefficient. We should write directly into the allocated array. The way this is architected needs rethinking.

Reply:
Yes, @xhochy and I discussed this above. The question is, do we want a few functions in, and then start iterating on a strategy (for ascii it's trivial, for utf8 less so), or not, I'm happy to explore different strategies first as well.

Reply:
I think we should start finding a sane strategy right now for the existing `ascii_upper` kernel. Indeed, a different strategy will be needed for utf8 `upper`.

In both cases, this probably means implementing those kernels as "vector" kernels, not "scalar" kernels (because the latter implies you only process one item at a time).

Reply:
For `ascii_upper`, the strategy should be simple:
* keep the same null bitmap buffer
* recompute a similar offsets buffer, but without the base offset
* allocate a same-size values buffer, and convert it in one pass (no need to iterate over individual array elements)

Reply:
Agree. Happy to take a look into turning this into a 'vector' kernel, if you have some pointer that would be great. This also answers a question I had if it would be possible to operate on a full array/chunk in the existing framework, I guess that's a yes.

Any reason not to reuse the offset buffer, even if the offset is not 0?

What is the advantage of scalar kernels in Arrow? Apart from avoiding boilerplate? Also wondering if this is related to the Gandiva connection.

Reply:
Avoiding boilerplate is the main motivation for scalar kernels. It really saves a lot of repetition in kernel implementations.

You can find a simple vector kernel example in `compute/kernels/vector_sort.cc`. Take a look especially at `RegisterVectorSort`, `AddSortingKernels` and `SortIndices`.

Reply:
I'm afraid you are mistaking what is meant by "ScalarFunction" and "VectorFunction". "Scalar" and "Vector" refer to the _semantics_ of the function, not the implementation. 

The only requirement for these string functions is that you provide `std::function<void(KernelContext*, const ExecBatch&, Datum*)>` that implements the kernel. What is inside can be anything. The function is **still** a ScalarFunction though